exist-core/src/main/java/org/exist/util/io/VirtualTempPath.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * eXist Open Source Native XML Database
+ * Copyright (C) 2001-2019 The eXist Project
+ * http://exist-db.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package org.exist.util.io;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.exist.util.FileUtils;
+
+/**
+ * @author Patrick Reinhart <patrick@reini.net>
+ */
+public final class VirtualTempPath implements AutoCloseable {
```

Comment:
I am wondering what the concurrency for this class is? Is it intended to be `@NotThreadSafe` or `@ThreadSafe`?

I note that the `content` member looks like it is designed to be thread-safe, but this class itself is not-thread-safe, due to its unsynchronized `inMemorySize`.

Reply:
Hmm, technically it is `@ThreadSafe`... Also how should the final field `inMemorySize` set within the Constructor be a problem? Did I miss something here?

Reply:
Sorry. `inMemorySize` is correct as it is `final`. But what about the visibility of `content` and `contentFile`? They are lazily initialised and not as an an atomic unit. If I am not mistaken, they need to be atomic references.